examples/research_projects/codeparrot/scripts/minhash_deduplication.py

```diff
@@ -0,0 +1,207 @@
+import json
+import multiprocessing as mp
+import re
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple, Type
+
+from datasets import Dataset
+from tqdm import tqdm
+
+from datasketch import MinHash, MinHashLSH
+from dpu_utils.utils.iterators import ThreadedIterator
+
+
+NUM_PERM = 256
+NON_ALPHA = re.compile("[^A-Za-z_0-9]")
+MIN_NUM_TOKENS = 10
+
+
+def get_min_hash(tokens: List[str]) -> Optional[MinHash]:
+    if len(tokens) < MIN_NUM_TOKENS:
+        return None
+    min_hash = MinHash(num_perm=NUM_PERM)
+    for token in set(tokens):
+        min_hash.update(token.encode())
+    return min_hash
+
+
+def get_tokens(code: str) -> Set[str]:
+    """
+    Tokenize a code snippet.
+    """
+    return set([t for t in NON_ALPHA.split(code) if len(t.strip()) > 0])
+
+
+class DuplicationIndex:
+    def __init__(
+        self,
+        *,
+        duplication_jaccard_threshold: float = 0.85,
+        num_perm: int = 256,
+        min_num_tokens: int = 10,
+    ):
+        self.__duplication_jaccard_threshold = duplication_jaccard_threshold
+        self.__num_perm = num_perm
+        self.__min_num_tokens = min_num_tokens
+        self.__index = MinHashLSH(threshold=self.__duplication_jaccard_threshold, num_perm=self.__num_perm)
+
+        self.__duplicate_clusters = defaultdict(set)
+
+    def add(self, filename: str, min_hash: MinHash) -> None:
+        close_duplicates = self.__index.query(min_hash)
+        if filename in self.__index.keys:
+            print("Duplicate key %s" % filename)
+            return
+
+        self.__index.insert(filename, min_hash)
+        if len(close_duplicates) > 0:
+            # print("`%s` duplicate of: %s" % (filename, close_duplicates))
+
+            for base_duplicate in close_duplicates:
+                if base_duplicate in self.__duplicate_clusters:
+                    self.__duplicate_clusters[base_duplicate].add(filename)
+                    break
+            else:
+                self.__duplicate_clusters[close_duplicates[0]].add(filename)
+
+    def get_duplicate_clusters(self) -> List[List[Dict]]:
+        duplicate_clusters = []
+        for base, duplicates in self.__duplicate_clusters.items():
+            cluster = [base] + list(duplicates)
+            # reformat the cluster to be a list of dict
+            cluster = [{"base_index": el[0], "repo_name": el[1], "path": el[2]} for el in cluster]
```

Comment:
Maybe we can also keep the `size` information if it doesn't slow the process, could be interesting to see how large are the duplicated files and how much data we remove (ofc the latter can also be computed after the filtering)

Reply:
Do you mean the size of the piece of code?

Reply:
yes I guess we have it in the original data